opm/common/utility/CSRGraphFromCoordinates.hpp

```diff
@@ -0,0 +1,491 @@
+/*
+  Copyright 2016 SINTEF ICT, Applied Mathematics.
+  Copyright 2016 Statoil ASA.
+  Copyright 2022 Equinor ASA
+
+  This file is part of the Open Porous Media Project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_UTILITY_CSRGRAPHFROMCOORDINATES_HPP
+#define OPM_UTILITY_CSRGRAPHFROMCOORDINATES_HPP
+
+#include <cstddef>
+#include <optional>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+/// \file
+///
+/// Facility for converting collection of region ID pairs into a sparse
+/// (CSR) adjacency matrix representation of a graph.  Supports O(nnz)
+/// compression.
+
+namespace Opm { namespace utility {
+
+    /// Form CSR adjacency matrix representation of unstructured graphs.
+    /// Optionally maps vertex pairs to compressed indices to enable O(1)
+    /// per-element lookup in assembly-like operations.
+    ///
+    /// \tparam VertexID ID type of an abstract vertex in the graph.  Could
+    ///    for instance be used to represent a cell index or a region index.
+    ///    Must be an integral type.
+    ///
+    /// \tparam TrackCompressedIdx Whether or not to form a mapping relation
+    ///    for vertex pairs to compressed indices.  Default value, false,
+    ///    bypasses this mapping relation and conserves memory.
+    template <typename VertexID = int, bool TrackCompressedIdx = false>
+    class CSRGraphFromCoordinates
```

Comment:
The "FromCoordinates" in the name confuses me, what are the coordinates in question?

Reply:
> what are the coordinates in question?

That refers to the "coordinate format (COO)" of sparse matrices.  That format is essentially a (long) list of (I,J) or (I,J,V) tuples.